crates/astria-conductor/src/reader.rs

```diff
@@ -282,3 +351,249 @@ impl Reader {
         Ok(())
     }
 }
+
+/// This function ensures that the given Commit has quorum, ie that the Commit contains >2/3 voting
+/// power. It performs the following checks:
+/// - the height of the commit matches the block height of the validator set
+/// - each validator in the commit is in the validator set
+/// - for each signature in the commit, the validator public key matches the validator address in
+///   the commit
+/// - for each signature in the commit, the validator signature in the commit is valid
+/// - the total voting power of the commit is >2/3 of the total voting power of the validator set
+///
+/// # Errors
+///
+/// If any of the above conditions are not satisfied, an error is returned.
+#[instrument]
+fn ensure_commit_has_quorum(
+    commit: &Commit,
+    validator_set: &ValidatorSet,
+    chain_id: &str,
+) -> Result<()> {
+    if commit.height != validator_set.block_height {
+        bail!(
+            "commit height mismatch: expected {}, got {}",
+            validator_set.block_height,
+            commit.height
+        );
+    }
+
+    let mut total_voting_power = 0u64;
+    validator_set.validators.iter().for_each(|v| {
+        total_voting_power += v.voting_power.0;
+    });
+
+    let validator_map = validator_set
+        .validators
+        .iter()
+        .map(|v| (&v.address, v)) // address is in bech32
+        .collect::<HashMap<_, _>>();
+
+    let mut commit_voting_power = 0u64;
+    for vote in &commit.signatures {
+        // we only care about votes that are for the Commit.BlockId (ignore absent validators and
+        // votes for nil)
+        if vote.block_id_flag != "BLOCK_ID_FLAG_COMMIT" {
+            continue;
+        }
+
+        // verify validator exists in validator set
+        let validator_address = bech32::encode(
+            "metrovalcons",
+            vote.validator_address.0.to_base32(),
+            Variant::Bech32,
+        )?;
+        let Some(validator) = validator_map.get(&validator_address) else {
+            bail!("validator {} not found in validator set", validator_address);
+        };
+
+        // verify address in signature matches validator pubkey
+        let address_from_pubkey = public_key_to_address(&validator.pub_key.key.0)?;
+        ensure!(
+            address_from_pubkey == validator_address,
+            format!(
+                "validator address mismatch: expected {}, got {}",
+                validator_address, address_from_pubkey
+            )
+        );
+
+        // verify vote signature
+        verify_vote_signature(
+            vote,
+            commit,
+            chain_id,
+            &validator.pub_key.key.0,
+            &vote.signature.0,
+        )?;
+
+        commit_voting_power += validator.voting_power.0;
+    }
+
+    ensure!(
+        commit_voting_power > total_voting_power * 2 / 3,
```

Comment:
This has overflow potential. Prefer `total_voting_power / 3 * 2`.

I would love to have this as a free standing function and a bunch of unit tests just for it, i.e.:
```rust
fn does_commited_voting_power_have_quorum(commited: u64, total: u64) -> bool {
    commited > total / 3 * 2
}

Reply:
unit tests added!